File: solutions-in-python/E-Knapsack.py
import sys

INF = (10**9)+5

n, w = map(int, input().split())
weight = []*n
value = []*n
for i in range(n):
	wt, val = map(int, input().split())
	weight.append(wt)
	value.append(val)

sum_value = 0
for x in value:
	sum_value += x

dp = [INF]*(sum_value+1)	
dp[0] = 0

for i in range(n):
	for value_cal in range(sum_value-value[i], -1, -1):
		dp[value_cal+value[i]] = min(dp[value_cal+value[i]], dp[value_cal]+weight[i]) 

ans = 0
for i in range(sum_value+1):
	if dp[i] <= w:
		ans = max(ans, i)

print(ans)			

File: solutions-in-cpp/E-Knapsack.cpp
#include<iostream>
#include<vector>

using namespace std;

#define ll long long

const int INF = 1e9+5;

int main(){
    //freopen("tree.in","r",stdin);
    //freopen("tree.out","w",stdout);
    //freopen("input.txt", "r", stdin);
    //freopen("out.txt", "w", stdout);
    ios_base::sync_with_stdio(0);
    //cin.tie(0);

    
    int n, w;
    cin >> n >> w;
    vector<ll> weight(n), value(n);

    for(int i = 0; i < n; ++i){
        cin >> weight[i] >> value[i];
    }
    int sum_value = 0;
    for(int i = 0; i < n; ++i){
        sum_value += value[i];
    }

    vector<ll> dp(sum_value+1, INF);
    dp[0] = 0;
    // dp[i] - the minimum total weight of items with total value exactly i
    for(int i = 0; i < n; ++i){
        for(int value_cal = sum_value-value[i]; value_cal >= 0; --value_cal){
            dp[value_cal+value[i]] = min(dp[value_cal+value[i]], dp[value_cal]+weight[i]);
            //cout << value_cal << " " << value[i] << " " << dp[value_cal+value[i]] << endl;
        }
    }

    ll ans = 0;
    for(int i = 0; i <= sum_value; ++i){
        if(dp[i] <= w){
            ans = max(ans, (ll)i);
            //cout << dp[i] << " " << ans << endl;
        }
    }
    cout << ans << endl;
}